ssl/ech/ech_store.c

```diff
@@ -0,0 +1,1168 @@
+        goto err;
+    }
+    switch (ee->version) {
+    case OSSL_ECH_RFCXXXX_VERSION:
+        break;
+    default:
+        /* skip over in case we get something we can handle later */
+        skip = OPENSSL_malloc(ech_content_length);
+        if (skip == NULL)
+            goto err;
+        if (!PACKET_copy_bytes(pkt, skip, ech_content_length)) {
+            ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+            OPENSSL_free(skip);
+            goto err;
+        }
+        OPENSSL_free(skip);
+        /* nothing to return but not a fail */
+        ossl_echstore_entry_free(ee);
+        return 1;
+    }
+    if (!PACKET_copy_bytes(pkt, &ee->config_id, 1)
+        || !PACKET_get_net_2(pkt, &tmpi)
+        || !PACKET_get_length_prefixed_2(pkt, &pub_pkt)
+        || (ee->pub_len = PACKET_remaining(&pub_pkt)) == 0
+        || (ee->pub = OPENSSL_malloc(ee->pub_len)) == NULL
+        || !PACKET_copy_bytes(&pub_pkt, ee->pub, ee->pub_len)
+        || !PACKET_get_length_prefixed_2(pkt, &cipher_suites)
+        || (suiteoctets = PACKET_remaining(&cipher_suites)) <= 0
+        || (suiteoctets % 2) == 1) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    thiskemid = (uint16_t) tmpi;
+    ee->nsuites = suiteoctets / OSSL_ECH_CIPHER_LEN;
+    ee->suites = OPENSSL_malloc(ee->nsuites * sizeof(*ee->suites));
+    if (ee->suites == NULL)
+        goto err;
+    while (PACKET_copy_bytes(&cipher_suites, cipher,
+                             OSSL_ECH_CIPHER_LEN)) {
+        ee->suites[ci].kem_id = thiskemid;
+        ee->suites[ci].kdf_id = cipher[0] << 8 | cipher [1];
+        ee->suites[ci].aead_id = cipher[2] << 8 | cipher [3];
+        if (ci++ > ee->nsuites) {
+            ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+    }
+    if (PACKET_remaining(&cipher_suites) > 0
+        || !PACKET_copy_bytes(pkt, &max_name_len, 1)) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    ee->max_name_length = max_name_len;
+    if (!PACKET_get_length_prefixed_1(pkt, &public_name_pkt)) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    public_name_len = PACKET_remaining(&public_name_pkt);
+    if (public_name_len < TLSEXT_MINLEN_host_name
+        || public_name_len > TLSEXT_MAXLEN_host_name) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    ee->public_name = OPENSSL_malloc(public_name_len + 1);
+    if (ee->public_name == NULL)
+        goto err;
+    if (PACKET_copy_bytes(&public_name_pkt,
+                          (unsigned char *) ee->public_name,
+                          public_name_len) != 1) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    ee->public_name[public_name_len] = '\0';
+    if (strlen(ee->public_name) != public_name_len) { /* no NUL inside! */
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
```

Comment:
Isn't it just this (replacing these 14 lines of code with 5)?
```suggestion
    if (PACKET_contains_zero_byte(&public_name_pkt)
            || !PACKET_strndup(&public_name_pkt, &(ee->public_name)) {
        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
        goto err;
    }
```

Reply:
actually, yep, you're right with the zero_byte check so changed to the suggestion.